### DISCUSSION TITLE
Can Files Be Assumed to be UTF-8?

### DISCUSSION THREAD

**Discussion by gavinhoward** [jgm/djot#78] [Q&A] [answered]
I'm preparing to write a djot library in C (for my own purposes and to implement an API for post-processing accessible from any language with bindings).

Can my library assume that djot files are UTF-8? Or will it be necessary to have conversions?

**Comment by matklad** [comment]
cc https://github.com/jgm/djot/issues/60, https://github.com/matklad/djot-rs

>to implement an API for post-processing accessible from any language with bindings

Another avenue for that is to go via canonical JSON AST https://github.com/jgm/djot/issues/58 (what I do today is shelling out to lua impl to get json, and then shove that into TypeScript which does the actual rendering like this:

<details>

```TypeScript
const visitor: { [key: string]: (node: Node) => HtmlString } = {
  doc: (node) => node.contents,
  heading: (node) => {
    const tag = `h${node.ast.level}`;
    const res = html`<${tag}>${node.contents}</${tag}>`;
    if (node.ast.level == 1) res.push(time(node.ctx.date));
    return res;
  },
  list: (node) => {
    const tag = "-+*".includes(node.ast.list_style) ? "ul" : "ol";
    return html`<${tag} ${node.class_attr}>${node.contents}</${tag}>`;
  },
  list_item: (node) => html`<li>${node.contents}</li>`,
  para: (node) => html`<p>${node.contents}</p>`,
  div: (node) => {
    if (node.cls.includes("note")) {
      return html`<aside class="note">
        <i class="fa icon-note"></i>
        <div>${node.contents}</div>
      </aside>`;
    }
    return html`<div class="${node.cls}">${node.contents}</div>`;
  },
  code_block: (node) => highlight(node.text, node.ast.lang),
  raw_block: (node) => {
    if (node.ast.format == "html") return new HtmlString(node.text);
    return html``;
  },
  table: (node) => html`<table>${node.contents}</table>`,
  row: (node) => html`<tr>${node.contents}</tr>`,
  cell: (node) => html`<td>${node.contents}</td>`,
  verbatim: (node) => html`<code>${node.text}</code>`,
  link: (node) => html`<a href="${node.ast.destination}">${node.text}</a>`,
  url: (node) =>
    html`<a class="url" href="${node.ast.destination}">${node.ast.destination}</a>`,
  double_quoted: (node) => html`“${node.contents}”`,
  emph: (node) => html`<em>${node.contents}</em>`,
  strong: (node) => html`<strong>${node.contents}</strong>`,
  thematic_break: (_node) => html`<br />`,
  str: (node) => html`${node.text}`,
};
```

</details>

> Can my library assume that djot files are UTF-8? Or will it be necessary to have conversions?

I think it should be OK for conforming impl to assume utf8; The spec itself I think isn't terribly sensitive to the encoding, we can spec it in terms of Unicode codepoints. We use byte-offsets for internal data structures, but I think thouse are not a part that should be covered by the spec.

**Reply by gavinhoward** [reply]
> cc https://github.com/jgm/djot/issues/60, https://github.com/matklad/djot-rs

Thank you, but I need something in C. I don't know Rust enough. However, I'll adopt your bug-for-bug compatibility with the Lua implementation.

> Another avenue for that is to go via canonical JSON AST https://github.com/jgm/djot/issues/58

I didn't want to implement JSON in C, but this is such a good idea that I will. I especially like that such an AST could allow a reader to be separate from a renderer, which would also allow extending of rendering. This could replace my uses of Pandoc for generating manpages, though Dr. MacFarlane may not like that idea.

> We use byte-offsets for internal data structures, but I think thouse are not a part that should be covered by the spec.

I agree with this.

**Comment by jgm** [comment, marked as answer]
> djot library in C

That sounds great!  But note, it is already possible to write a C bridge that links to liblua and just passes in the djot source code as a string. `make djotbin` will already do this, essentially, to create a statically linked binary.  A similar approach could be used to create a C library that would always be in sync with the Lua implementation.  (It wouldn't be as fast as a native C library, but since the Lua implementation is already plenty fast, this may not be an issue.)

> Can my library assume that djot files are UTF-8?

I agree, I don't think it's necessary to specify a particular encoding in the spec, but as noted the Lua implementation just assumes UTF-8.

This is something that could be discussed further, though.  Perhaps it would be worth enforcing UTF-8, but to me this seems a concrete implementation detail rather than something about the markup syntax.

**Comment by gavinhoward** [comment]
> But note, it is already possible to write a C bridge that links to liblua and just passes in the djot source code as a string. `make djotbin` will already do this, essentially, to create a statically linked binary. A similar approach could be used to create a C library that would always be in sync with the Lua implementation. (It wouldn't be as fast as a native C library, but since the Lua implementation is already plenty fast, this may not be an issue.)

This is a good idea, and if I knew Lua and its C interop, I would be willing to contribute such a bridge to your implementation.

But I'm picky about dependencies, so I will still write my own implementation from scratch. I also think having another would help with momentum (https://github.com/jgm/djot/issues/60#issuecomment-1304920065) and for testing out the language. I don't think it will be too much work.

> I agree, I don't think it's necessary to specify a particular encoding in the spec, but as noted the Lua implementation just assumes UTF-8.

Okay, thank you.

> This is something that could be discussed further, though. Perhaps it would be worth enforcing UTF-8, but to me this seems a concrete implementation detail rather than something about the markup syntax.

Because the encoding could be changed before ever hitting a reader, I think requiring UTF-8 is a good idea. Re-encoding can happen beforehand, especially if implementations allow "passing in the djot source code as a string."

So perhaps djot can be encoded in anything, but implementations are only expected to read it in UTF-8?

**Comment by jgm** [comment]
> I'm picky about dependencies, so I will still write my own implementation from scratch.

FWIW, the dependencies required to build-in Lua are quite low.  You just need liblua.a and some header files.  It's about 360K.  Whether or not you create an independent C library, I think it would be worth someone doing this.

> So perhaps djot can be encoded in anything, but implementations are only expected to read it in UTF-8?
 
I'm thinking this can be up to the implementation. An implementation can say, "we presuppose that the input is UTF-8 encoded."

**Comment by gavinhoward** [comment]
> FWIW, the dependencies required to build-in Lua are quite low. You just need liblua.a and some header files. It's about 360K.

Understood.

> Whether or not you create an independent C library, I think it would be worth someone doing this.

I think so too, and I would do it if I knew how. I would even use it in testing my independent library. Maybe I'll learn Lua's C interop and submit a PR to this repo.

> I'm thinking this can be up to the implementation. An implementation can say, "we presuppose that the input is UTF-8 encoded."

I think this is a good way to go.

**Reply by jgm** [reply]
I've added a `clib` directory with an initial experiment in this direction.
It creates a library libdjot.a that exports a function that will convert a string to a JSON AST.
